### PR TITLE
feat(lib): add safe wrappers for PMIx_Envar/Setenv/Multicluster_nspace

### LINKED ISSUE #75: [Missing Safe Wrapper] PMIx_Multicluster_nspace_parse
The PMIx function `PMIx_Multicluster_nspace_parse` does not appear to have a safe Rust wrapper in the pmix-rs crate.

**Function:** `PMIx_Multicluster_nspace_parse`
**Suggested Module:** `unknown` (based on naming convention)

**Description:**
This function is present in the PMIx C API bindings but lacks a corresponding safe Rust wrapper. Safe wrappers should be added to provide idiomatic Rust interface to PMIx functionality.

**Steps to Implement:**
1. Add a safe wrapper function in the appropriate module (likely `src/unknown.rs`)
2. Follow the existing patterns in the codebase for safe wrappers
3. Ensure proper error handling and return types
4. Add documentation and examples if appropriate
5. Consider adding tests

**Reference:**
- See existing safe wrappers in modules like `data_ops.rs`, `events.rs`, `process_mgmt.rs`, etc.
- The function should typically take ownership of resources where appropriate and return `PmixResult<T>` or similar.

This issue was automatically generated as part of a sweep to identify missing safe wrappers in the pmix-rs crate.

### LINKED ISSUE #95: [Missing Safe Wrapper] PMIx_Multicluster_nspace_parse
The PMIx function `PMIx_Multicluster_nspace_parse` does not appear to have a safe Rust wrapper in the pmix-rs crate.

**Function:** `PMIx_Multicluster_nspace_parse`
**Suggested Module:** `unknown` (based on naming convention)

**Description:**
This function is present in the PMIx C API bindings but lacks a corresponding safe Rust wrapper. Safe wrappers should be added to provide idiomatic Rust interface to PMIx functionality.

**Steps to Implement:**
1. Add a safe wrapper function in the appropriate module (likely `src/unknown.rs`)
2. Follow the existing patterns in the codebase for safe wrappers
3. Ensure proper error handling and return types
4. Add documentation and examples if appropriate
5. Consider adding tests

**Reference:**
- See existing safe wrappers in modules like `data_ops.rs`, `events.rs`, `process_mgmt.rs`, etc.
- The function should typically take ownership of resources where appropriate and return `PmixResult<T>` or similar.

This issue was automatically generated as part of a sweep to identify missing safe wrappers in the pmix-rs crate.

### LINKED ISSUE #142: [Missing Safe Wrapper] PMIx_Setenv
The PMIx function `PMIx_Setenv` does not appear to have a safe Rust wrapper in the pmix-rs crate.

**Function:** `PMIx_Setenv`
**Suggested Module:** `unknown` (based on naming convention)

**Description:**
This function is present in the PMIx C API bindings but lacks a corresponding safe Rust wrapper. Safe wrappers should be added to provide idiomatic Rust interface to PMIx functionality.

**Steps to Implement:**
1. Add a safe wrapper function in the appropriate module (likely `src/unknown.rs`)
2. Follow the existing patterns in the codebase for safe wrappers
3. Ensure proper error handling and return types
4. Add documentation and examples if appropriate
5. Consider adding tests

**Reference:**
- See existing safe wrappers in modules like `data_ops.rs`, `events.rs`, `process_mgmt.rs`, etc.
- The function should typically take ownership of resources where appropriate and return `PmixResult<T>` or similar.

This issue was automatically generated as part of a sweep to identify missing safe wrappers in the pmix-rs crate.

### LINKED ISSUE #151: [Missing Safe Wrapper] PMIx_Envar_free
The PMIx function `PMIx_Envar_free` does not appear to have a safe Rust wrapper in the pmix-rs crate.

**Function:** `PMIx_Envar_free`
**Suggested Module:** `unknown` (based on naming convention)

**Description:**
This function is present in the PMIx C API bindings but lacks a corresponding safe Rust wrapper. Safe wrappers should be added to provide idiomatic Rust interface to PMIx functionality.

**Steps to Implement:**
1. Add a safe wrapper function in the appropriate module (likely `src/unknown.rs`)
2. Follow the existing patterns in the codebase for safe wrappers
3. Ensure proper error handling and return types
4. Add documentation and examples if appropriate
5. Consider adding tests

**Reference:**
- See existing safe wrappers in modules like `data_ops.rs`, `events.rs`, `process_mgmt.rs`, etc.
- The function should typically take ownership of resources where appropriate and return `PmixResult<T>` or similar.

This issue was automatically generated as part of a sweep to identify missing safe wrappers in the pmix-rs crate.

### LINKED ISSUE #154: [Missing Safe Wrapper] PMIx_Envar_destruct
The PMIx function `PMIx_Envar_destruct` does not appear to have a safe Rust wrapper in the pmix-rs crate.

**Function:** `PMIx_Envar_destruct`
**Suggested Module:** `unknown` (based on naming convention)

**Description:**
This function is present in the PMIx C API bindings but lacks a corresponding safe Rust wrapper. Safe wrappers should be added to provide idiomatic Rust interface to PMIx functionality.

**Steps to Implement:**
1. Add a safe wrapper function in the appropriate module (likely `src/unknown.rs`)
2. Follow the existing patterns in the codebase for safe wrappers
3. Ensure proper error handling and return types
4. Add documentation and examples if appropriate
5. Consider adding tests

**Reference:**
- See existing safe wrappers in modules like `data_ops.rs`, `events.rs`, `process_mgmt.rs`, etc.
- The function should typically take ownership of resources where appropriate and return `PmixResult<T>` or similar.

This issue was automatically generated as part of a sweep to identify missing safe wrappers in the pmix-rs crate.

### LINKED ISSUE #158: [Missing Safe Wrapper] PMIx_Envar_construct
The PMIx function `PMIx_Envar_construct` does not appear to have a safe Rust wrapper in the pmix-rs crate.

**Function:** `PMIx_Envar_construct`
**Suggested Module:** `unknown` (based on naming convention)

**Description:**
This function is present in the PMIx C API bindings but lacks a corresponding safe Rust wrapper. Safe wrappers should be added to provide idiomatic Rust interface to PMIx functionality.

**Steps to Implement:**
1. Add a safe wrapper function in the appropriate module (likely `src/unknown.rs`)
2. Follow the existing patterns in the codebase for safe wrappers
3. Ensure proper error handling and return types
4. Add documentation and examples if appropriate
5. Consider adding tests

**Reference:**
- See existing safe wrappers in modules like `data_ops.rs`, `events.rs`, `process_mgmt.rs`, etc.
- The function should typically take ownership of resources where appropriate and return `PmixResult<T>` or similar.

This issue was automatically generated as part of a sweep to identify missing safe wrappers in the pmix-rs crate.

### LINKED ISSUE #160: [Missing Safe Wrapper] PMIx_Multicluster_nspace_construct
The PMIx function `PMIx_Multicluster_nspace_construct` does not appear to have a safe Rust wrapper in the pmix-rs crate.

**Function:** `PMIx_Multicluster_nspace_construct`
**Suggested Module:** `unknown` (based on naming convention)

**Description:**
This function is present in the PMIx C API bindings but lacks a corresponding safe Rust wrapper. Safe wrappers should be added to provide idiomatic Rust interface to PMIx functionality.

**Steps to Implement:**
1. Add a safe wrapper function in the appropriate module (likely `src/unknown.rs`)
2. Follow the existing patterns in the codebase for safe wrappers
3. Ensure proper error handling and return types
4. Add documentation and examples if appropriate
5. Consider adding tests

**Reference:**
- See existing safe wrappers in modules like `data_ops.rs`, `events.rs`, `process_mgmt.rs`, etc.
- The function should typically take ownership of resources where appropriate and return `PmixResult<T>` or similar.

This issue was automatically generated as part of a sweep to identify missing safe wrappers in the pmix-rs crate.

### LINKED ISSUE #195: [Missing Safe Wrapper] PMIx_Envar_load
The PMIx function `PMIx_Envar_load` does not appear to have a safe Rust wrapper in the pmix-rs crate.

**Function:** `PMIx_Envar_load`
**Suggested Module:** `unknown` (based on naming convention)

**Description:**
This function is present in the PMIx C API bindings but lacks a corresponding safe Rust wrapper. Safe wrappers should be added to provide idiomatic Rust interface to PMIx functionality.

**Steps to Implement:**
1. Add a safe wrapper function in the appropriate module (likely `src/unknown.rs`)
2. Follow the existing patterns in the codebase for safe wrappers
3. Ensure proper error handling and return types
4. Add documentation and examples if appropriate
5. Consider adding tests

**Reference:**
- See existing safe wrappers in modules like `data_ops.rs`, `events.rs`, `process_mgmt.rs`, etc.
- The function should typically take ownership of resources where appropriate and return `PmixResult<T>` or similar.

This issue was automatically generated as part of a sweep to identify missing safe wrappers in the pmix-rs crate.

### LINKED ISSUE #311: [Missing Safe Wrapper] PMIx_Envar_create
The PMIx function `PMIx_Envar_create` does not appear to have a safe Rust wrapper in the pmix-rs crate.

**Function:** `PMIx_Envar_create`
**Suggested Module:** `unknown` (based on naming convention)

**Description:**
This function is present in the PMIx C API bindings but lacks a corresponding safe Rust wrapper. Safe wrappers should be added to provide idiomatic Rust interface to PMIx functionality.

**Steps to Implement:**
1. Add a safe wrapper function in the appropriate module (likely `src/unknown.rs`)
2. Follow the existing patterns in the codebase for safe wrappers
3. Ensure proper error handling and return types
4. Add documentation and examples if appropriate
5. Consider adding tests

**Reference:**
- See existing safe wrappers in modules like `data_ops.rs`, `events.rs`, `process_mgmt.rs`, etc.
- The function should typically take ownership of resources where appropriate and return `PmixResult<T>` or similar.

This issue was automatically generated as part of a sweep to identify missing safe wrappers in the pmix-rs crate.

### PR DESCRIPTION
**Pull request by SedahsDev**
Closes #75, Closes #95, Closes #142, Closes #151, Closes #154, Closes #158, Closes #160, Closes #195, Closes #311

## Summary
- Add envar construct/destruct/load wrappers and PmixEnvarArray (create/free) in src/lib.rs (home of the real PmixEnvar; orphan files untouched).
- Add setenv() Vec<String>-based wrapper.
- Add multicluster_nspace_construct/parse wrappers.
- Append mock FFI for daemon-free tests.
## Test plan
- Local mock-based tests; daemon/DVM tests remain CI-only.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -5742,3 +5742,203 @@ mod value_utils_tests {
...
+    );
+    pmix_ffi_or_mock!(
+        mock = unsafe { mock_ffi::mock_envar_destruct(&mut raw) },
+        real = unsafe { ffi::PMIx_Envar_destruct(&mut raw) },
+    );
+    Ok(result)
+}
+
+/// Invoke the PMIx destructor contract without exposing the raw struct.
+pub fn envar_destruct(_envar: &mut PmixEnvar) {
```
Comment (SedahsReview) [inline src/lib.rs]: **Blocking:** This function ignores `_envar` and calls `PMIx_Envar_destruct` on a newly zeroed temporary. It therefore never destructs the caller-provided `PmixEnvar` (nor frees its CString fields), making the public destructor a no-op. Please either retain/construct a raw `pmix_envar_t` that owns the same allocations and destruct that object, or remove/reshape this API so its ownership semantics are real.

File: src/lib.rs [hunk 2]
```diff
@@ -5742,3 +5742,203 @@ mod value_utils_tests {
...
+
+fn c_string_vector(values: &[String]) -> Result<*mut *mut c_char, PmixStatus> {
+    let slots = values.len().checked_add(1).ok_or(PmixStatus::from_raw(PmixError::ErrBadParam as i32))?;
+    let argv = unsafe { libc::calloc(slots, mem::size_of::<*mut c_char>()) as *mut *mut c_char };
+    if argv.is_null() { return Err(PmixStatus::from_raw(PmixError::ErrNomem as i32)); }
+    for (index, value) in values.iter().enumerate() {
+        let c = match CString::new(value.as_str()) {
+            Ok(c) => c,
+            Err(_) => {
+                unsafe { libc::free(argv.cast()) };
```
Comment (SedahsReview) [inline src/lib.rs]: **Blocking:** On an interior-NUL in a later `values` entry, `argv` is freed but the earlier `CString::into_raw()` pointers stored in slots are leaked. This is reachable with ordinary `Vec<String>` input and repeats on every failure. Free each initialized string (or use an RAII argv builder) before freeing the slot array.

**Review by SedahsReview — CHANGES_REQUESTED** [review]
## Verdict: REQUEST_CHANGES

The PR compiles and the existing unit suite passes, but the new public wrappers contain correctness and memory-safety issues that should be fixed before merge. In particular, `envar_destruct` does not operate on the supplied value, and the `setenv` preparation path leaks already-created strings when a later input contains an interior NUL. The new tests currently use no-op mocks for the critical FFI behavior, so they do not catch either issue or exercise the multicluster length boundary.

### Gate results
- `cargo build`: PASS
- `cargo test --lib`: PASS (1812 passed, 29 ignored; 82 existing warnings)
- `cargo test --lib envar_utils_tests`: PASS (5 passed)
- `cargo check --features mock_ffi --lib`: PASS (44 warnings, matching the main baseline)
- `cargo clippy --lib -- -D warnings`: PASS
- Diff: append-only, `src/lib.rs` +200 / `src/mock_ffi.rs` +10

### Blocking issues
- `src/lib.rs:5822`: `envar_destruct` ignores the caller's `PmixEnvar` and destructs a fresh zeroed C object, so it cannot release the strings owned by the supplied value and the public destructor is semantically a no-op. The wrapper needs an ownership model that maps the Rust value to the raw object passed to PMIx, or it should not expose this operation as a destructor.
- `src/lib.rs:5877`: if any later environment string contains an interior NUL, the error path frees only the pointer array and leaks every prior `CString::into_raw()` allocation. Free the already-created elements (or use an RAII argv builder) before freeing the slots.

### Suggestions
- Add behaviorful mocks/tests for `PMIx_Setenv` that replace `*env` and verify the returned Vec is copied from the updated array and both old/new allocations are released.
- Add tests for `envar_create(0)`, NUL inputs, and multicluster inputs at/over the PMIX_MAX_NSLEN budget. The current multicluster tests only validate no-op mock behavior.

**Comment by SedahsReview** [comment]
## Code Review Summary

**Verdict: REQUEST_CHANGES**

### Critical
- **src/lib.rs:5822** — `envar_destruct` ignores the supplied `PmixEnvar` and destructs a fresh zeroed object, so it does not release the value being passed by the caller.
- **src/lib.rs:5877** — interior-NUL failure in `c_string_vector` leaks every previously created `CString::into_raw()` allocation.

### Warnings
- The mock implementations are no-ops for the important pointer-update/free behavior, so the tests do not validate `PMIx_Setenv` reallocation or multicluster output/length handling.

### Gate results
- `cargo build`: PASS
- `cargo test --lib`: PASS (1812 passed, 29 ignored)
- focused `envar_utils_tests`: PASS (5 passed)
- `cargo check --features mock_ffi --lib`: PASS (44 warnings)
- `cargo clippy --lib -- -D warnings`: PASS
- diff is append-only: 200 insertions in `src/lib.rs`, 10 in `src/mock_ffi.rs`

Please address the two blocking issues and add behaviorful tests around allocator ownership and the PMIX_MAX_NSLEN boundary.

**Review by SedahsReview — APPROVED** [review]
## Verdict: APPROVE

Re-reviewed fix commits `65ecba0` and `f740a5f` at `f740a5f3ccd600f42a35898e72ce209054a0d5a7`. The two blocking findings from review #4889019911 are addressed, and the requested mock/test coverage is present. No new blocking issues found in the changed code.

### Verified fixed
- `envar_destruct` no longer passes a temporary raw object to PMIx. It is explicitly a no-op because `PmixEnvar` owns Rust `CString` fields; the documentation states that PMIx must not be called on those Rust-owned allocations, and normal Rust destruction releases them without freeing Rust-owned memory through PMIx.
- `c_string_vector` now frees every initialized string before freeing the slot array when a later `CString::new` fails on an interior NUL.
- The Setenv mock now models appending/replacement, copies the environment into a replacement array when needed, and frees the old array/entries. Tests verify the resulting Vec and the later interior-NUL error path.
- Added tests cover `envar_create(0)`, behaviorful multicluster mocks, and rejection of oversized namespace input.

### Gates
- `cargo build`: PASS
- `cargo test --lib`: PASS (1816 passed, 29 ignored)
- `cargo test --lib envar_utils_tests`: PASS (9 passed)
- `cargo check --features mock_ffi --lib`: PASS (44 warnings, matching the stated baseline)
- `cargo clippy --lib -- -D warnings`: PASS
- `git diff --check`: PASS
- PR diff remains append-only: `src/lib.rs` and `src/mock_ffi.rs` only (+288 lines, no deletions against the PR base).

**Review by SedahsReview — APPROVED** [review]
## Verdict: APPROVE

Re-review of fix commits `65ecba0` and `f740a5f` at `f740a5f3ccd600f42a35898e72ce209054a0d5a7`. Both blocking findings from review #4889019911 are addressed.

### Verified fixed
- `envar_destruct` now documents and enforces the Rust-owned `CString` model by intentionally making no PMIx call; Rust destruction releases the fields without PMIx freeing Rust-owned allocations.
- `c_string_vector` frees each initialized string before freeing the pointer array on later interior-NUL failure.
- The Setenv mock models append/replacement and frees old allocations; tests verify replacement and interior-NUL behavior.
- Tests cover `envar_create(0)`, behaviorful multicluster mocks, and oversized namespace input.

### Gates
- `cargo build`: PASS
- `cargo test --lib`: PASS (1816 passed, 29 ignored)
- `cargo test --lib envar_utils_tests`: PASS (9 passed)
- `cargo check --features mock_ffi --lib`: PASS (44 warnings, baseline)
- `cargo clippy --lib -- -D warnings`: PASS
- `git diff --check`: PASS
- PR diff is append-only against the PR base: 2 changed files, +288 lines.

**Comment by SedahsReview** [comment]
## Review summary

**Verdict: APPROVE**

Verified fixes for review #4889019911: `envar_destruct` now has explicit Rust-owned CString semantics and performs no unsafe PMIx destructor call; `c_string_vector` frees all initialized strings on interior-NUL failure; Setenv and multicluster mocks are behaviorful; and tests cover zero-length envar arrays, replacement/error paths, and namespace length validation.

Gate results: `cargo build` PASS; `cargo test --lib` PASS (1816 passed, 29 ignored); `cargo test --lib envar_utils_tests` PASS (9 passed); `cargo check --features mock_ffi --lib` PASS (44 baseline warnings); `cargo clippy --lib -- -D warnings` PASS; `git diff --check` PASS. The PR diff remains append-only against the PR base.

Formal review: #4889044234 (APPROVE).
